File: src/core/clean-executor.ts
/**
 * Clean executor module for ubuild
 *
 * Executes clean operations for Unreal Engine projects by removing build
 * artifacts including Binaries, Intermediate, and Saved directories.
 * Supports dry-run mode and selective cleaning.
 *
 * @module core/clean-executor
 */

import fs from 'fs-extra';
import path from 'path';
import { Writable } from 'stream';
import { CleanOptions, CleanResult } from '../types/clean';
import { formatError } from '../utils/error';
import { Logger } from '../utils/logger';
import { ProjectPathResolver } from './project-path-resolver';

/**
 * Executes clean operations for Unreal Engine projects.
 * Removes build artifacts including Binaries, Intermediate, and Saved directories.
 */
export class CleanExecutor {
  private logger: Logger;
  private stdout: Writable;
  private stderr: Writable;
  private silent: boolean;

  /**
   * Creates a new CleanExecutor instance.
   * @param options - Configuration options for logging and output streams
   */
  constructor(
    options: { logger?: Logger; stdout?: Writable; stderr?: Writable; silent?: boolean } = {}
  ) {
    this.stdout = options.stdout || process.stdout;
    this.stderr = options.stderr || process.stderr;
    this.silent = options.silent || false;
    this.logger =
      options.logger ||
      new Logger({
        stdout: this.stdout,
        stderr: this.stderr,
        silent: this.silent,
      });
  }

  /**
   * Executes a clean operation with the specified options.
   * Removes build artifacts from the project directory.
   * @param options - Clean configuration options
   * @returns Promise resolving to clean result with success status and deleted paths
   */
  async execute(options: CleanOptions): Promise<CleanResult> {
    const deletedPaths: string[] = [];
    const failedPaths: Array<{ path: string; error: string }> = [];

    try {
      const projectPath = await ProjectPathResolver.resolveOrThrow(
        options.projectPath || process.cwd()
      );

      const projectDir = path.dirname(projectPath);
      const projectName = path.basename(projectPath, '.uproject');

      // Define paths to clean
      const pathsToClean = this.getPathsToClean(projectDir, projectName, options.binariesOnly);

      if (!this.silent) {
        this.logger.info(`Cleaning project: ${projectName}`);
        this.logger.info(`Project directory: ${projectDir}`);

        if (options.binariesOnly) {
          this.logger.info('Mode: Binaries and Intermediate only');
        } else {
          this.logger.info('Mode: Full clean (Binaries, Intermediate, Saved, DerivedDataCache)');
        }

        if (options.dryRun) {
          this.logger.info('Dry run mode - no files will be deleted');
        }

        this.logger.divider();
      }

      // Clean each path
      for (const cleanPath of pathsToClean) {
        const result = await this.cleanPath(cleanPath, options.dryRun);
        if (result.deleted) {
          deletedPaths.push(cleanPath);
          if (!this.silent) {
            this.logger.success(`Removed: ${path.relative(projectDir, cleanPath)}`);
          }
        } else if (result.error) {
          failedPaths.push({ path: cleanPath, error: result.error });
          if (!this.silent) {
            this.logger.error(`Failed to remove: ${path.relative(projectDir, cleanPath)}`);
          }
        }
      }

      // Clean plugin directories
      const pluginResults = await this.cleanPluginDirectories(projectDir, options);
      deletedPaths.push(...pluginResults.deletedPaths);
      failedPaths.push(...pluginResults.failedPaths);

      // Check for failures BEFORE the silent block — success must not depend on logging mode
      if (failedPaths.length > 0) {
        if (!this.silent) {
          this.logger.divider();
        }
        return {
          success: false,
          deletedPaths,
          failedPaths,
          error: `Failed to clean ${failedPaths.length} path(s)`,
        };
      }

      if (!this.silent) {
        this.logger.divider();

        if (deletedPaths.length === 0) {
          this.logger.info('No build artifacts found to clean');
        } else {
          this.logger.success(`Cleaned ${deletedPaths.length} directory/directories`);
        }
      }

      return {
        success: true,
        deletedPaths,
        failedPaths,
      };
    } catch (error) {
      return {
        success: false,
        deletedPaths,
        failedPaths,
        error: formatError(error),
      };
    }
  }

  /**
   * Gets the list of paths to clean based on the clean mode.
   * @param projectDir - The project directory path
   * @param projectName - The name of the project
   * @param binariesOnly - Whether to clean only Binaries and Intermediate
   * @returns Array of absolute paths to clean
   */
  private getPathsToClean(
    projectDir: string,
    projectName: string,
    binariesOnly?: boolean
  ): string[] {
    const paths: string[] = [
      path.join(projectDir, 'Binaries'),
      path.join(projectDir, 'Intermediate'),
    ];

    if (!binariesOnly) {
      paths.push(
        path.join(projectDir, 'Saved'),
        path.join(projectDir, 'DerivedDataCache'),
        path.join(projectDir, `${projectName}.sln`),
        path.join(projectDir, '.vs'),
        path.join(projectDir, '.idea')
      );
    }

    return paths;
  }

  /**
   * Cleans a single path (file or directory).
   * @param cleanPath - The path to clean
   * @param dryRun - Whether to perform a dry run
   * @returns Object indicating whether the path was deleted or an error occurred
   */
  private async cleanPath(
    cleanPath: string,
    dryRun?: boolean
  ): Promise<{ deleted?: boolean; error?: string }> {
    try {
      const exists = await fs.pathExists(cleanPath);
      if (!exists) {
        return { deleted: false };
      }

      if (!dryRun) {
        await fs.remove(cleanPath);
      }

      return { deleted: true };
    } catch (error) {
      return {
        deleted: false,
        error: formatError(error),
      };
    }
  }

  /**
   * Cleans build directories from plugin folders.
   * @param projectDir - The project directory path
   * @param options - Clean options
   * @returns Object containing arrays of deleted and failed paths
   */
  private async cleanPluginDirectories(
    projectDir: string,
    options: CleanOptions
  ): Promise<{ deletedPaths: string[]; failedPaths: Array<{ path: string; error: string }> }> {
    const deletedPaths: string[] = [];
    const failedPaths: Array<{ path: string; error: string }> = [];

    const pluginsDir = path.join(projectDir, 'Plugins');
    const pluginsExist = await fs.pathExists(pluginsDir);

    if (!pluginsExist) {
      return { deletedPaths, failedPaths };
    }

    try {
      const pluginDirs = await fs.readdir(pluginsDir);

      for (const pluginDir of pluginDirs) {
        const fullPluginPath = path.join(pluginsDir, pluginDir);
        const stat = await fs.stat(fullPluginPath);

        if (!stat.isDirectory()) {
          continue;
        }

        // Clean Binaries and Intermediate in each plugin
        const pluginBinaries = path.join(fullPluginPath, 'Binaries');
        const pluginIntermediate = path.join(fullPluginPath, 'Intermediate');

        for (const cleanPath of [pluginBinaries, pluginIntermediate]) {
          const result = await this.cleanPath(cleanPath, options.dryRun);
          if (result.deleted) {
            deletedPaths.push(cleanPath);
            if (!this.silent) {
              this.logger.success(`Removed: ${path.relative(projectDir, cleanPath)}`);
            }
          } else if (result.error) {
            failedPaths.push({ path: cleanPath, error: result.error });
            if (!this.silent) {
              this.logger.error(`Failed to remove: ${path.relative(projectDir, cleanPath)}`);
            }
          }
        }
      }
    } catch (error) {
      const errorMsg = formatError(error);
      if (!this.silent) {
        this.logger.warning(`Failed to clean plugins: ${errorMsg}`);
      }
      failedPaths.push({ path: pluginsDir, error: errorMsg });
    }

    return { deletedPaths, failedPaths };
  }
}

File: src/core/clean_executor.rs
use std::fs;
use std::path::{Path, PathBuf};

use anyhow::Result;

use crate::types::{CleanResult, PROJECT_OPERATION_STATE_DIR};
use crate::utils::logger::Logger;

use super::engine_resolver::EngineResolver;
use super::project_path_resolver::ProjectPathResolver;

pub struct CleanExecutor;

impl CleanExecutor {
    pub fn execute(
        project: Option<&str>,
        engine_path: Option<&str>,
        dry_run: bool,
        binaries_only: bool,
    ) -> Result<CleanResult> {
        let (project_path, _engine) =
            EngineResolver::resolve_project_and_engine(project, engine_path)?;

        let project_name = ProjectPathResolver::project_name(&project_path);
        let project_dir = ProjectPathResolver::project_dir(&project_path);
        let saved_dir = project_dir.join("Saved");

        Logger::info(&format!("Cleaning project: {project_name}"));
        Logger::info(&format!("Project directory: {}", project_dir.display()));
        Logger::info(&format!(
            "Mode: {}",
            if binaries_only {
                "Binaries and Intermediate only"
            } else {
                "Full clean"
            }
        ));
        if dry_run {
            Logger::info("Dry run mode - no files will be deleted");
        }
        Logger::divider();

        let mut paths_to_clean: Vec<PathBuf> = vec![
            project_dir.join("Binaries"),
            project_dir.join("Intermediate"),
        ];

        if !binaries_only {
            paths_to_clean.extend([
                saved_dir.clone(),
                project_dir.join("DerivedDataCache"),
                project_dir.join(format!("{project_name}.sln")),
                project_dir.join(".vs"),
                project_dir.join(".idea"),
            ]);
        }

        let mut result = CleanResult {
            deleted_paths: Vec::new(),
            failed_paths: Vec::new(),
        };

        for path in &paths_to_clean {
            if path == &saved_dir {
                Self::clean_saved(path, &project_dir, dry_run, &mut result);
            } else {
                Self::clean_path(path, &project_dir, dry_run, &mut result);
            }
        }

        // Clean plugin directories
        let plugins_dir = project_dir.join("Plugins");
        if plugins_dir.is_dir() {
            if let Ok(entries) = fs::read_dir(&plugins_dir) {
                for entry in entries.filter_map(Result::ok) {
                    let plugin_dir = entry.path();
                    if plugin_dir.is_dir() {
                        Self::clean_path(
                            &plugin_dir.join("Binaries"),
                            &project_dir,
                            dry_run,
                            &mut result,
                        );
                        Self::clean_path(
                            &plugin_dir.join("Intermediate"),
                            &project_dir,
                            dry_run,
                            &mut result,
                        );
                    }
                }
            }
        }

        Logger::divider();
        let count = result.deleted_paths.len();
        if count > 0 {
            Logger::success(&format!("Cleaned {count} item(s)"));
        } else {
            Logger::info("Nothing to clean");
        }

        Ok(result)
    }

    fn clean_saved(saved_dir: &Path, project_dir: &Path, dry_run: bool, result: &mut CleanResult) {
        let state_dir = saved_dir.join(PROJECT_OPERATION_STATE_DIR);
        if !state_dir.exists() {
            Self::clean_path(saved_dir, project_dir, dry_run, result);
            return;
        }

        let Ok(entries) = fs::read_dir(saved_dir) else {
            Self::clean_path(saved_dir, project_dir, dry_run, result);
            return;
        };
        for entry in entries.filter_map(Result::ok) {
            let path = entry.path();
            if path == state_dir {
                continue;
            }
            Self::clean_path(&path, project_dir, dry_run, result);
        }
    }

    fn clean_path(path: &Path, project_dir: &Path, dry_run: bool, result: &mut CleanResult) {
        if !path.exists() {
            return;
        }

        let relative = path
            .strip_prefix(project_dir)
            .unwrap_or(path)
            .display()
            .to_string();

        if dry_run {
            Logger::success(&format!("Would remove: {relative}"));
            result.deleted_paths.push(path.to_path_buf());
            return;
        }

        let remove_result = if path.is_dir() {
            fs::remove_dir_all(path)
        } else {
            fs::remove_file(path)
        };

        match remove_result {
            Ok(()) => {
                Logger::success(&format!("Removed: {relative}"));
                result.deleted_paths.push(path.to_path_buf());
            }
            Err(e) => {
                Logger::error(&format!("Failed to remove {relative}: {e}"));
                result
                    .failed_paths
                    .push((path.to_path_buf(), e.to_string()));
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use std::fs;

    use anyhow::Result;
    use tempfile::tempdir;

    use super::CleanExecutor;
    use crate::types::{CleanResult, PROJECT_OPERATION_STATE_DIR};

    #[test]
    fn full_clean_preserves_project_operation_state() -> Result<()> {
        let project = tempdir()?;
        let saved = project.path().join("Saved");
        let state = saved.join(PROJECT_OPERATION_STATE_DIR);
        let logs = saved.join("Logs");
        fs::create_dir_all(&state)?;
        fs::create_dir_all(&logs)?;
        fs::write(state.join("operation.lock"), "")?;
        fs::write(logs.join("editor.log"), "log")?;
        fs::write(saved.join("temporary.txt"), "temporary")?;
        let mut result = CleanResult {
            deleted_paths: Vec::new(),
            failed_paths: Vec::new(),
        };

        CleanExecutor::clean_saved(&saved, project.path(), false, &mut result);

        assert!(state.join("operation.lock").is_file());
        assert!(!logs.exists());
        assert!(!saved.join("temporary.txt").exists());
        assert!(result.failed_paths.is_empty());
        Ok(())
    }
}
